src/helpers/shutdown.ts

```diff
@@ -0,0 +1,55 @@
+import Logger from './Logger'
+import { Command } from 'types'
+
+const log = new Logger('helpers:shutdown')
+type QuitSignal = 'SIGINT' | 'SIGTERM' | 'SIGQUIT'
+const POSIX_SIGNALS: QuitSignal[] = ['SIGINT', 'SIGTERM', 'SIGQUIT']
+const listeners: Command[] = []
+let isRunning = true
+
+POSIX_SIGNALS.forEach(signal => {
+  process.on(signal, () => {
+    _doShutDown(`I've gotten a ${signal} signal`)
+  })
+})
+
+export function onShutdown (listener: Command) {
```

Comment:
I don't understand - `onShutdown` you're pushing _new_ listeners into the array and then calling them (`line 31`) where are you removing the `listeners` e.g empting the array?

Reply:
Exactly. But line 31 is part of the exported `shutDown`.

So there's 2 ways two stop the app:
* **Declarative**: You call the exported shutDown interface
* **Sigterm**: You receive a sigterm signal (ctrl+c) or systemd restart, or when kubernetes wants to restart the app, ...

The `onShutdown` allows to subscribe to the event, to allow to do some clean up. In this branch no cleanup is done. It's just saying bye. But probably we need to cleanup the WS of web3 or telegram.